lib/phoenix/socket.ex

```diff
@@ -122,50 +123,285 @@ defmodule Phoenix.Socket do
+    %{
+      endpoint: endpoint,
+      serializer: serializer,
+      transport: transport,
+      params: params
+    } = map
+
+    # The information in the Phoenix.Socket goes to userland and channels.
+    socket = %Socket{
+      handler: handler,
+      endpoint: endpoint,
+      pubsub_server: endpoint.__pubsub_server__,
+      serializer: serializer,
+      transport: transport
+    }
+
+    # The information in the state is kept only inside the socket process.
+    state = %{
+      channels: %{},
+      channels_inverse: %{}
+    }
+
+    case handler.connect(params, socket) do
+      {:ok, %Socket{} = socket} ->
+        case handler.id(socket) do
+          nil ->
+            {:ok, {state, socket}}
+
+          id when is_binary(id) ->
+            {:ok, {state, %{socket | id: id}}}
+
+          invalid ->
+            Logger.error "#{inspect handler}.id/1 returned invalid identifier #{inspect invalid}. " <>
+                         "Expected nil or a string."
+            :error
+        end
+
+      :error ->
+        :error
+
+      invalid ->
+        Logger.error "#{inspect handler}.connect/2 returned invalid value #{inspect invalid}. " <>
+                     "Expected {:ok, socket} or :error"
+        :error
+    end
+  end
+
+  def __init__({state, %{id: id, endpoint: endpoint} = socket}) do
+    _ = id && endpoint.subscribe(id, link: true)
+    {:ok, {state, %{socket | transport_pid: self()}}}
+  end
+
+  def __in__({payload, opts}, {state, socket}) do
+    %{topic: topic} = message = socket.serializer.decode!(payload, opts)
+    handle_in(Map.get(state.channels, topic), message, state, socket)
+  end
+
+  def __info__({:DOWN, ref, _, pid, reason}, {state, socket}) do
+    case state.channels_inverse do
+      %{^pid => {topic, join_ref}} ->
+        state = delete_channel(state, pid, topic, ref)
+        {:push, encode_on_exit(socket, topic, join_ref, reason), {state, socket}}
+
+      %{} ->
+        {:ok, {state, socket}}
+    end
+  end
+
+  def __info__({:graceful_exit, pid, %Phoenix.Socket.Message{} = message}, {state, socket}) do
+    state =
+      case state.channels_inverse do
+        %{^pid => {topic, _join_ref}} ->
+          {^pid, monitor_ref} = Map.fetch!(state.channels, topic)
+          delete_channel(state, pid, topic, monitor_ref)
+
+        %{} ->
+          state
+      end
+
+    {:push, encode_reply(socket, message), {state, socket}}
+  end
+
+  def __info__(%Broadcast{event: "disconnect"}, state) do
+    {:stop, {:shutdown, :disconnected}, state}
+  end
+
+  def __info__({:socket_push, opcode, payload}, state) do
+    {:push, {opcode, payload}, state}
+  end
+
+  def __info__(:garbage_collect, state) do
+    :erlang.garbage_collect(self())
+    {:ok, state}
+  end
+
+  def __info__(_, state) do
+    {:ok, state}
+  end
+
+  def __terminate__(_reason, {%{channels_inverse: channels_inverse}, _socket}) do
+    Phoenix.Channel.Server.close(Map.keys(channels_inverse))
```

Comment:
Wouldn't it make sense to exit the supervisor for the channels here instead of manually closing the connections?

Reply:
The channel supervisor is shared across all sockets.

Reply:
We could change it so each socket has its own supervisor but that would imply even more sockets per connection. The opposite concern though is that the supervisor we added may become a bottleneck. Unlikely though.